[DEMO-2502] Add mysql on-host newrelic integration

moonlight-komorebi opened the pull request:
* Added the tasks & template for installing the integration, as well as
verifying that it works once installed. The integration installation
uses the public new relic infra ansible role.
* The role is 'onafterstart' because this needs mysql to be installed.
The installation of mysql occurs in 'start', so this must follow that.

moonlight-komorebi committed:
[DEMO-2502] Add mysql on-host newrelic integration

* Added the tasks & template for installing the integration, as well as
verifying that it works once installed. The integration installation
uses the public new relic infra ansible role.
* The role is 'onafterstart' because this needs mysql to be installed.
The installation of mysql occurs in 'start', so this must follow that.

Julien4218 commented on deploy/databases/linux/mysql/roles/onafterstart/tasks/main.yml:
```diff
@@ -0,0 +1,74 @@
+    password: "{{ database_password }}"
+    state: present
+    host: "localhost"
+    priv: "*.*:GRANT,REPLICATION CLIENT"
```
I believe it also needs the SELECT permission, despite what our NR documentation says (I have a feedback item to the infra OHI team about that).

moonlight-komorebi replied:
do you know why this is needed? the integration appears to be working

Julien4218 replied:
On some version of linux/mysql, this was required. I believed Debian10 with latest mysql version

Julien4218 commented on deploy/databases/linux/mysql/roles/onafterstart/tasks/main.yml:
```diff
@@ -0,0 +1,74 @@
+- name: Create db user for newrelic mysql integration
+  mysql_user:
+    login_user: "root"
+    login_password: "{{ database_password }}"
```
I think we should have a different password for the newrelic user and the root password. We shouldn't assume or make those 2 credentials share a same password.

moonlight-komorebi replied:
updated so users have different passwords, with the newrelic user password generated within the play.

Julien4218 commented on deploy/databases/linux/mysql/roles/onafterstart/tasks/main.yml:
```diff
@@ -0,0 +1,74 @@
+    msg: "newrelic_license_key is required for configuring newrelic infrastructure agent"
+  when: newrelic_license_key is not defined
+
+- name: Ensure database_port is defined
```
Maybe the port could be optional?  Defaulting to 3306.

Julien4218 commented on deploy/databases/linux/mysql/roles/onafterstart/tasks/main.yml:
```diff
@@ -0,0 +1,74 @@
+      log_file: /var/log/nr-infra.log
+      log_to_stdout: false
+    nrinfragent_integrations:
+      - { name: nri-mysql, state: "latest" }
```
Normally, all instrumentation definitions support a version field. I'm inferring here we're only installing latest.
Is there a way to pin a specific version, as we typically do on all other newrelic instrumentation plays?

Julien4218 commented on deploy/databases/linux/mysql/roles/onafterstart/tasks/main.yml:
```diff
@@ -0,0 +1,74 @@
+  become: true
+
+# Give the agent time to start up & write to log
+- name: Wait 15 seconds
```
Instead of an absolute wait, we could do a loop/until, so we can try multiple times at some interval to check for the file to be written.
Maybe every 5 seconds, up to a minute.

Julien4218 commented on deploy/databases/linux/mysql/roles/onafterstart/tasks/templates/mysql-config.yml:
```diff
@@ -0,0 +1,28 @@
+        extended_metrics: 1
+        extended_innodb_metrics: 1
+        extended_myisam_metrics: 1
+    labels:
```
I dont think those labels below are correct.
Instead we should set all the labels/tags that are defined and pass through from the deployer.

moonlight-komorebi replied:
added tags as labels within the play.

Julien4218 reviewed (changes requested):
A good start, I've put a few feedback items to look into.

moonlight-komorebi committed:
[DEMO-2502] Make port optional

* Added a new port variable which is effectively optional. It uses the
pre-existing variable if its defined, otherwise defaults to 3306.

moonlight-komorebi committed:
[DEMO-2502] Generate newrelic user password

* Remove the sharing of passwords and instead generate the password as
part of the play. The password generation is idempotent, so multiple
deployments use the first password generated.

moonlight-komorebi committed:
[DEMO-2502] Remove existing labels, add tags as labels

* Removed existing labels
* Added a step to the play to add tags as labels

moonlight-komorebi committed:
[DEMO-2502] Swap out hard wait with loop until

* Instead of waiting for a specific amount of time, use loop until.

moonlight-komorebi committed:
[DEMo-2502] Add SELECT permission to newrelic integration user

* Adding SELECT permission to the newrelic integration user based on PR
comment that this is needed for certain setups.

moonlight-komorebi committed:
[DEMO-2502] Move clear logs step

* Moving the step that clears logs after the agent role. This allows us
to remove the stat step, since the log file will exist.
